DSharpPlus.Interactivity/InteractivityExtension.cs

```diff
@@ -115,12 +118,63 @@ public async Task<ReadOnlyCollection<PollEmoji>> DoPollAsync(DiscordMessage m, I
         }
 
         /// <summary>
-        /// Waits for a user to submit a modal.
+        /// Waits for a modal with the specified id to be submitted.
         /// </summary>
-        /// <param name="modal_id">The id of the modal to wait for.</param>
+        /// <param name="modal_id">The id of the modal to wait for. Should be unique to avoid issues.</param>
         /// <param name="timeoutOverride">Override the timeout period in <see cref="InteractivityConfiguration"/>.</param>
+        /// <returns>A <see cref="InteractivityResult{ModalSubmitEventArgs}"/> with a modal if the interactivity did not time out.</returns>
         public Task<InteractivityResult<ModalSubmitEventArgs>> WaitForModalAsync(string modal_id, TimeSpan? timeoutOverride = null)
-            => this.WaitForEventArgsAsync<ModalSubmitEventArgs>(m => m.Interaction.Data.CustomId == modal_id, timeoutOverride);
+            => this.WaitForModalAsync(modal_id, this.GetCancellationToken(timeoutOverride));
+
+        /// <summary>
+        /// Waits for a modal with the specified id to be submitted.
+        /// </summary>
+        /// <param name="modal_id">The id of the modal to wait for. Should be unique to avoid issues.</param>
+        /// <param name="token">A custom cancellation token that can be cancelled at any point.</param>
+        /// <returns>A <see cref="InteractivityResult{ModalSubmitEventArgs}"/> with a modal if the interactivity did not time out.</returns>
+        public async Task<InteractivityResult<ModalSubmitEventArgs>> WaitForModalAsync(string modal_id, CancellationToken token)
+        {
+            if (string.IsNullOrEmpty(modal_id) || modal_id.Length > 100)
+                throw new ArgumentException("Custom ID must be between 1 and 100 characters.");
+
+            var matchRequest = new ModalMatchRequest(modal_id,
+                    c => c.Interaction.Type == InteractionType.ModalSubmit &&
```

Comment:
If this is hooked to ModalSubmitted, this check is redundant

Reply:
The check is redundant in the sense that any submitted modal will have a correct id, but it's still a runtime check for making sure the user doesn't wait for an incorrect custom id. WaitForButtonAsync() methods have similar runtime checks to check a message for components, though they don't check for correct id's. 

Would you and/or @OoLunar like me to remove it?

Reply:
If it's redundant, then I'd remove it **after** checking for no behavioral differences. This would involve testing a before and after.

Reply:
I can tell you this is redundant because I wrote the code and this event doesn't fire if it's not a modal /lh

Reply:
https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus/Clients/DiscordClient.Dispatch.cs#L2515-L2520

Reply:
if ModalSubmitted fired when the data wasn't a modal we'd have far bigger issues to fix than interactivity 👍🏽

Reply:
Ah, sorry. Github on desktop made it seem like you were referring to line 138, the check for custom ID rather than the check for InteractionType.

Then yes, the check is redundant as ModalSubmitted only fires with a modal.